Ensure we register a cycle if we've got scala sources but pipelining is disabled

ephemerist opened the pull request.

ephemerist committed:
Ensure we register a cycle if we've got scala sources but pipelining is disabled

ephemerist commented:
THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE Organization Contribution License Agreement v2.0, dated July 13, 2012.THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

ephemerist commented:
Getting a 500 error trying to sign the Scala CLA

lrytz commented:
> Getting a 500 error trying to sign the Scala CLA

I reported it

ephemerist commented:
> > Getting a 500 error trying to sign the Scala CLA
> 
> I reported it

Thanks. Scala CLA now complete.

eed3si9n commented:
Thanks for the contribution. While this is fresh to you, could you retroactively create an issue describing the bug you're fixing here?

ephemerist commented:
> Thanks for the contribution. While this is fresh to you, could you retroactively create an issue describing the bug you're fixing here?

Added as #1512 - let me know if you'd like any more detail.

eed3si9n commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -70,14 +70,15 @@ object Incremental {
      * and decide whether we need another cycle.
      */
-    def mergeAndInvalidate(partialAnalysis: Analysis, completingCycle: Boolean): CompileCycleResult
+    def mergeAndInvalidate(partialAnalysis: Analysis, registerCycle: Boolean): CompileCycleResult
```
nit: consider renaming this to something like `shouldRegisterCycle` since `def registerCycle` exists, which this flag is encouraging the implementer to later call.

eed3si9n commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -70,14 +70,15 @@ object Incremental {
         prev: Option[CompileCycleResult],
-        partialAnalysis: Analysis
+        partialAnalysis: Analysis,
+        registerCycle: Boolean
```
nit: same here.

eed3si9n commented on internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala:
```diff
@@ -1090,12 +1097,12 @@ private final class AnalysisCallback(
         val a = getAnalysis
         val CompileCycleResult(continue, invalidations, merged) =
-          incHandler.mergeAndInvalidate(a, false)
+          incHandler.mergeAndInvalidate(a, true)
```
to avoid boolean blindness, could we add `parameterName = true` here plz?

ephemerist commented:
@eed3si9n thanks for the review - I've addressed all your comments. Not sure why Scala CLA is still failing - I clicked the link and it says `{"user":"ephemerist","signed":true,"version":"1.0","currentVersion":"1.0"}`

lrytz commented:
> Not sure why Scala CLA is still failing

Seems the checker doesn't follow redirects. https://github.com/sbt/zinc/pull/1513

eed3si9n committed:
Merge branch '1.10.x' into feature/cycles-registration

eed3si9n reviewed (approved):
Thanks!
